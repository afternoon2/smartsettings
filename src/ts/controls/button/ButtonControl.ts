import { Control } from '../control/Control';
import {
  ButtonControlProps, ControlListener,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/button.sass';

export class ButtonControl extends Control {
  public controlElement: HTMLButtonElement;

  protected static defaultType: string = 'button';

  protected static template = (state: InternalState): string => `
    <button
      class="${Styles.button}"
      type="${state.type || ButtonControl.defaultType}"
      title="${state.name}"
      ${state.disabled ? 'disabled="true"' : ''}
    >
      ${state.text || ''}
    </button>
  `;

  private controlListener?: ControlListener;
  private sectionListener?: ControlListener;
  private panelListener?: ControlListener;

  constructor(props: ButtonControlProps) {
    super(props, ButtonControl.template);
    this.controlElement = this.element.querySelector('button') as HTMLButtonElement;
    this.controlListener = props.listener;
    this.sectionListener = props.sectionListener;
    this.panelListener = props.panelListener;
    this.bindActionListeners();
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('click', () => {
      if (this.controlListener) {
        this.controlListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'control'
        });
      }
      if (this.sectionListener) {
        this.sectionListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'section',
        });
      }
      if (this.panelListener) {
        this.panelListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'panel'
        });
      }
    });
  }
}
