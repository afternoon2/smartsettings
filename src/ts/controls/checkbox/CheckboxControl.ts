import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, CheckboxControlOptions } from '../../types';

import Styles from '../../../sass/checkbox.sass';

export type CheckboxControlProps = {
  id: string,
  options: CheckboxControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class CheckboxControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'checkbox';

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
      class="${Styles.checkbox}" for="${id}">
        <input
          id="${id}"
          type="checkbox"
          title="${state.name}"
          ${state.checked ? 'checked' : ''}
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `;
  }

  constructor(props: CheckboxControlProps) {
    super({
      ...props,
      template: CheckboxControl.template,
    });
    this.controlElement = this.element.querySelector('[type="checkbox"]') as HTMLInputElement;
    this.listeners.set('checked', this.onCheck);
    this.bindActionListeners();
  }

  get checked(): boolean {
    return this.state.checked as boolean;
  }

  check() {
    this.state.checked = true;
  }

  uncheck() {
    this.state.checked = false;
  }

  toggle() {
    this.state.checked = !this.state.checked;
  }

  private onCheck: Listener = (update: ListenerUpdate) => {
    if (update.value === true) {
      this.controlElement.setAttribute('checked', 'true');
    } else {
      this.controlElement.removeAttribute('checked');
    }
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      event.preventDefault();
      this.toggle();
    });
  }
}
