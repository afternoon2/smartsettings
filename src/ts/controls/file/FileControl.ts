import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, FileControlOptions } from '../../types';

import Styles from '../../../sass/file.sass';

export type FileControlProps = {
  id: string,
  options: FileControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class FileControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
        class="${Styles.file}"
        for="${id}"
      >
        Upload...
        <input
          id="${id}"
          type="file"
          title="${state.name}"
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `
  };

  constructor(props: FileControlProps) {
    super({
      ...props,
      template: FileControl.template,
    });
    this.controlElement = this.element.querySelector('[type="file"]') as HTMLInputElement;
    this.listeners.set('accept', this.onAcceptChange);
    this.bindActionListeners();
  }

  get accept(): string {
    return this.state.accept as string;
  }

  set accept(accept: string) {
    this.state.accept = accept;
  }

  get files(): FileList | null {
    return this.state.files as FileList;
  }

  private onAcceptChange: Listener = (update: ListenerUpdate) => {
    this.controlElement.setAttribute('accept', update.value as string);
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      this.state.files = ((event as Event).target as HTMLInputElement).files as FileList;
    });
  }
}
