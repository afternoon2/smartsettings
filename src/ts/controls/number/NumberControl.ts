import { Control } from '../control/Control';
import {
  NumberControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/number.sass';

export class NumberControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="number"
      class="${Styles.number}"
      ${state.value ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.disabled ? 'disabled' : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
    />
  `;

  constructor(props: NumberControlProps) {
    super(props, NumberControl.template);
    this.controlElement = this.element.querySelector('[type="number"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('change', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }
}