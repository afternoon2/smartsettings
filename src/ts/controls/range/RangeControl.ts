import { Control } from '../control/Control';
import {
  RangeControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/range.sass';
import ControlStyles from '../../../sass/control.sass';

export class RangeControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="range"
      class="${Styles.range}"
      ${state.value && typeof state.value === 'number' ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly' : ''}
      ${state.disabled ? 'disabled' : ''}
    />
  `;

  constructor(props: RangeControlProps) {
    super(props, RangeControl.template);
    this.controlElement = this.element.querySelector('[type="range"]') as HTMLInputElement;
    this.updateMain();
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
    (this.element.querySelector(`.${Styles.rangeValue}`) as HTMLSpanElement).textContent = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }

  private updateMain = () => {
    const main = this.element.querySelector(`.${ControlStyles.control__content}`) as HTMLElement;
    main.classList.add(`${ControlStyles['control__content--range']}`);
    main.insertAdjacentHTML('beforeend', `<span class="${Styles.rangeValue}">
      ${typeof this.state.value === 'number' ? this.state.value : ''}
    </span>`);
  }
}