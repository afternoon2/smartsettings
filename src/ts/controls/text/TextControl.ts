import { Control } from '../control/Control';
import { InternalState, TextControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/text.sass';

export type TextControlProps = {
  id: string,
  options: TextControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'text';

  protected static template = (state: InternalState): string => `
    <input
      type="text"
      class="${Styles.text}"
      ${state.disabled ? 'disabled' : ''}
      ${state.value ? `value="${state.value}"` : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `min-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.pattern ? `pattern="${state.pattern}"` : ''}
      ${state.readOnly ? 'readonly="true"' : ''}
    />
  `;

  constructor(props: TextControlProps) {
    super({
      ...props,
      template: TextControl.template
    });
    this.controlElement = this.element.querySelector('[type="text"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
