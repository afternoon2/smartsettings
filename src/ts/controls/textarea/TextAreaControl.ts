import { Control } from '../control/Control';
import { InternalState, TextAreaControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/textarea.sass';

export type TextAreaControlProps = {
  id: string,
  options: TextAreaControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextAreaControl extends Control {
  public controlElement: HTMLTextAreaElement;

  protected static defaultRows: number = 4;
  protected static defaultCols: number = 10;

  protected static template = (state: InternalState): string => `
    <textarea
      wrap="hard"
      class="${Styles.textarea}"
      cols="${state.cols || TextAreaControl.defaultCols}"
      rows="${state.rows || TextAreaControl.defaultRows}"
      ${state.disabled ? 'disabled="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `max-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
    >${state.value || ''}</textarea>
  `;

  constructor(props: TextAreaControlProps) {
    super({
      ...props,
      template: TextAreaControl.template
    });
    this.controlElement = this.element.querySelector('textarea') as HTMLTextAreaElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
