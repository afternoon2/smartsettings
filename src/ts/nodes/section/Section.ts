import { SectionProps } from '../nodes.types';
import { ParentNode } from '../parent/ParentNode';
import { InternalState } from '../../root/RootNode';

import Base from '../../../sass/base.sass';
import Styles from '../../../sass/section.sass';

export class SectionNode extends ParentNode {
  public headerElement: HTMLElement;
  public bodyElement: HTMLElement;

  protected static template = (state: InternalState): string => `<header
      class="${Styles.section__header}"
    >
      <p class="${Styles.section__name}">
        ${state.name}
      </p>
      <a class="${Styles.section__link}">
        ${state.collapsed ? '&#9662;' : '&#9656;'}
      </a>
    </header>
    <main class="${Styles.section__body} ${state.collapsed ? Base.hidden : ''}"></main>
  `;

  constructor(props: SectionProps) {
    super(props, SectionNode.template);
    this.fillInElement(Styles.section);
    this.headerElement = this.element.querySelector(`.${Styles.section__header}`) as HTMLElement;
    this.bodyElement = this.element.querySelector(`.${Styles.section__body}`) as HTMLElement;
    this.bindEventListeners();
  }

  private onHeaderClick = (event: Event) => {
    event.preventDefault();
    if (this.state.collapsed === true) {
      this.open();
    } else {
      this.close();
    }
  }

  private bindEventListeners() {
    this.headerElement.addEventListener('click', this.onHeaderClick);
  }
}