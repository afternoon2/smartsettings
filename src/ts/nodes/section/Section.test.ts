import { SectionNode } from './Section';
import { ButtonControl } from '../../controls/button/ButtonControl';
import { TextControl } from '../../controls/text/TextControl';
import { TextAreaControl } from '../../controls/textarea/TextAreaControl';
import { NumberControl } from '../../controls/number/NumberControl';
import { RangeControl } from '../../controls/range/RangeControl';
import { CheckboxControl } from '../../controls/checkbox/CheckboxControl';
import { FileControl } from '../../controls/file/FileControl';

let section: SectionNode;
let listener: any;

describe('Section node', () => {
  beforeEach(() => {
    listener = jest.fn();
    section = new SectionNode({
      id: 'id',
      name: 'name',
      options: {
        collapsed: false,
        disabled: false,
        invisible: false,
      },
      parentElement: document.body,
      listener,
    });
  });

  test('If it matches the snapshot', () => {
    expect(section.element).toMatchSnapshot();
  });

  test('Open and close', () => {
    section.close();
    expect(section.collapsed).toBe(true);
    expect(section.element).toMatchSnapshot();
    section.open();
    expect(section.collapsed).toBe(false);
    expect(listener).toHaveBeenCalledTimes(2);
    expect(section.element).toMatchSnapshot();
    section.toggle();
    expect(section.collapsed).toBe(true);
    expect(section.element).toMatchSnapshot();
  });

  test('Open and close by click', () => {
    (section.headerElement.querySelector('a') as HTMLAnchorElement).click();
    expect(section.collapsed).toBe(true);
    expect(section.element).toMatchSnapshot();
  });

  test('Control creation', () => {
    const btn = section.control('button', 'Button', {
      text: 'Click me!',
    });
    const text = section.control('text', 'Text', {
      placeholder: 'Write here!',
      value: 'text input',
    });
    const textarea = section.control('textarea', 'Textarea', {
      placeholder: 'textarea',
      value: 'Textarea',
    });
    const range = section.control('range', 'Range', {
      min: 0, max: 10, value: 1, step: 1,
    });
    const number = section.control('number', 'Range', {
      min: 0, max: 10, value: 1, step: 1,
    });
    const checkbox = section.control('checkbox', 'Checkbox', {
      checked: true,
    });
    const file = section.control('file', 'Checkbox', {});
    const falsyControl = section.control('falsy', 'Falsy', {
      min: 1, max: 10, step: 1, value: 10,
    });
    expect(btn).toBeInstanceOf(ButtonControl);
    expect(text).toBeInstanceOf(TextControl);
    expect(textarea).toBeInstanceOf(TextAreaControl);
    expect(number).toBeInstanceOf(NumberControl);
    expect(range).toBeInstanceOf(RangeControl);
    expect(checkbox).toBeInstanceOf(CheckboxControl);
    expect(file).toBeInstanceOf(FileControl);
    expect(falsyControl).toBe(null);
  });
});

