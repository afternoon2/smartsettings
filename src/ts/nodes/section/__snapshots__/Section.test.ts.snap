// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section node If it matches the snapshot 1`] = `
<div
  class="section"
>
  <header
    class="section__header"
  >
    
      
    <p
      class="section__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="section__link"
    >
      
        ▸
      
    </a>
    
    
  </header>
  
    
  <main
    class="section__body "
  />
  
  
</div>
`;

exports[`Section node Open and close 1`] = `
<div
  class="section"
>
  <header
    class="section__header"
  >
    
      
    <p
      class="section__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="section__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="section__body hidden"
  />
  
  
</div>
`;

exports[`Section node Open and close 2`] = `
<div
  class="section"
>
  <header
    class="section__header"
  >
    
      
    <p
      class="section__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="section__link"
    >
      ▸
    </a>
    
    
  </header>
  
    
  <main
    class="section__body"
  />
  
  
</div>
`;

exports[`Section node Open and close by click 1`] = `
<div
  class="section"
>
  <header
    class="section__header"
  >
    
      
    <p
      class="section__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="section__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="section__body hidden"
  />
  
  
</div>
`;
